### ISSUE TITLE
recover_message gives different address from EIP-712 signature

### ISSUE THREAD

**Issue by wbobeirne** [ethereum/eth-account#71]
### What was wrong?

I'm trying to sign some data using EIP-712 with their Web3 client (Using Metamask in my case), and then recover the address on the backend to verify that a user is actually the owner of that address. However, the address I get back from `Account.recover_message` doesn't match it.

#### Client JS
```ts
const chainId = await web3.eth.net.getId();
const [address] = await web3.eth.getAccounts();
const data = {
  types: {
    EIP712Domain: [
      { name: 'name', type: 'string' },
      { name: 'version', type: 'string' },
      { name: 'chainId', type: 'uint256' },
      { name: 'verifyingContract', type: 'string' },
    ],
    authorization: [
      { name: 'userid', type: 'uint256' },
      { name: 'point', type: 'string' },
    ],
  },
  domain: {
    name: 'test.test',
    version: '1',
    chainId,
    verifyingContract: '0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC',
  },
  primaryType: 'authorization',
  message: {
    userid: 123,
    point: 'abc',
  },
};

const signature = await web3.currentProvider.send('eth_signTypedData_v3', [
  address,
  JSON.stringify(data),
]);

// send signature and data to python backend...
```

#### Server Python
```py
encoded = encode_structured_data(primitive=data)
addr = Account.recover_message(encoded, signature=signature)
# python addr !== js address
```

---

Let me know if I'm not doing this correctly, or if there are known differences between Metamask's EIP-712 format and this library's.

**Comment by kclowes** [comment]
If I am reading your JS right, it looks like you're sending the data in as a string, but you're passing it to `encode_structured_data` with the primitive option. If you use `encode_structured_data(text=data)` do you still get different addresses?

**Comment by wbobeirne** [comment]
I'm deserializing the JSON string on my backend, it's passing the validation parts of the code and all that good stuff.

**Comment by wbobeirne** [comment]
I can also confirm I'm able to correctly recover the address using the npm module `eth-sig-util`'s [`recoverTypedSignature`](https://github.com/MetaMask/eth-sig-util/blob/master/index.js#L369) function with a nearly identical simple webserver setup. It's pretty hard to tell where I should start debugging this though.

**Comment by kclowes** [comment]
Yeah, I don't see anything obviously wrong. It's possible that the implementations are different, but I see tests in both repos that make sure the address is the same as the one that signed the message. Here is the line from the EIP's [Example.js](https://github.com/ethereum/eth-account/blob/master/tests/core/test_structured_data_signing.py#L170), and the corresponding [tests in eth-account](https://github.com/ethereum/eth-account/blob/master/tests/core/test_structured_data_signing.py#L170) 
and the [eth-sig-util repo](https://github.com/MetaMask/eth-sig-util/blob/2648e7632be7ce8fa21bd29729b6b4d22c96723a/test/index.js#L467).
Are you able to correctly recover the signature with this library without using the message passed from your front end, like you did with `eth-sig-util`?

**Comment by wbobeirne** [comment]
Yeah, I tried to reconstruct the message with the raw data, rather than passing the whole message in, and also couldn't seem to recover it correctly. I'll continue to tinker with it on my end and let you know if I can narrow it down, or at the very least come up with a good self-contained reproduction of the issue. Thanks for taking a look.

**Comment by pipermerriam** [comment]
> or at the very least come up with a good self-contained reproduction of the issue.

This would be helpful.

**Comment by wbobeirne** [comment]
Just an update, I've actually encountered the exact same issue that provides the exact same incorrect address for me in the latest version of `eth-sig-util`, outlined here: https://github.com/MetaMask/eth-sig-util/issues/56

I'm guessing the same fix that caused this issue is also found in this library. It's hard to say if it's MetaMask or the libraries that're doing something wrong here, but hopefully this helps track it down.

**Comment by Dirt-Nasty** [comment]
@wbobeirne Did you ever solve this?  I'm looking to do something similar to verify a user on a Python backend and I can't recover_message to return the correct address.

**Comment by wbobeirne** [comment]
We ended up using a node.js microservice that used `eth-sig-util` to do the verification, which worked fine. Sorry!

**Comment by tyrnism** [comment]
It's a late comment but the issue in your code was the 'verifyingContract' type: it's not 'string' but 'address'.

**Comment by fselmo** [comment]
> It's a late comment but the issue in your code was the 'verifyingContract' type: it's not 'string' but 'address'.

Closing due to staleness and the above comment. Actual differences with `eth-sig-utils`, for `signTypedData_V4` (current), should also be addressed in PR #175, though I don't believe that would've affected this example and the above comment is likely where the issue was.

Feel free to re-open if there are still issues here!

**Comment by nazariyv** [comment]
this might be useful: https://ethereum.stackexchange.com/questions/137491/inconsistency-in-recovered-account-when-verifying-eip-712-javascript-python/137569#137569
